### ISSUE TITLE
[Proposal] [ADR Included] [Design Feedback Requested] Suggestion Refinement for Advanced Installer

### ISSUE THREAD

**Issue by Sewer56** [Nexus-Mods/NexusMods.App#634]
# Advanced Installer: Improve Suggestions

Note: This is a backlogged iterative improvement to the Advanced Installer, not part of Min Viable Product for this feature.

Approx Time: 2-3 hours.
Implement the filtering logic described in the ADR below.

## Prerequisites

- https://github.com/Nexus-Mods/NexusMods.App/issues/633

## Checklist

Method signature should look something like `(string filePath, bool isFolder, ISuggestionFolder[] existing)`.

- [ ] Implement Directory Filter Algorithm
- [ ] Implement File Filter Algorithm

----------------------

# ADR BELOW

(To be updated accordingly and included in PR which implements this).
(Append the text to existing ADR for suggestions during PR/upon completion)

----------------------

# Advanced Installer: Suggestions Refinement

## Filtering Suggestions (Directories)

Note: String comparisons performed here adhere to ![Paths Doc](./0003-paths.md). [Case insensitive + / separator]

The suggestion algorithm for folders is described below.
We will use modified example from [Advanced Installer Design](./0009-advanced-installer-design.md).

Consider the following directory structure:

```text
Red Hair
├─ Data
│  ├─ model.dae
│  └─ hair.dds
Green Hair
├─ Data
│  ├─ model.dae
│  └─ hair.dds
Blue Hair
└─ Data
   ├─ model.dae
   └─ hair.dds
```

Assume the `Data` folder is a folder within the game directory (also applies to subfolders).

If the user selects `model.dae` in `Blue Hair`; the suggestion algorithm should try finding the following in the game's directory tree:
- `Blue Hair/Data`
- `Data`

We keep removing the first directory in the path, until there's at least 1 match in the game's directory tree
(substring sanitized path!). Once there is at least 1 match, include that in the suggestions and terminate the search.

tl;dr Include any folders where any `substring` of the in-archive directory path, matches an `InstallFolderTarget`.

## Filtering Suggestions (Files)

The filtering algorithm for suggestions where to place a file functions in the following manner:

- Filter by `KnownValidFileExtensions`. (Remove if unknown extension).
- Filter by `FileExtensionsToDiscard`. (Remove if extension to discard).
- Add any directories from [Implementation Algorithm (Directories)](#implementation-algorithm-directories)

------------

Feedback from design ( @captainsandypants ) is requested; some suggestions of mine are below.

## How we Use the Filtered Data

- Sort such that most likely folders appear at the top of the list (in current design).
  - In other words, the items left after filtering are placed first.

or

-  Separate current 'Suggestions' box into 'Suggested' and 'Other Common Folders'.
  - i.e. Put files that left after filtering 1 box, and all other common locations in another box.
